system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1314,411 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+                {
+                  dir_in_use = TRUE;
+                  break;
+                }
+            }
+          G_UNLOCK (ongoing_pulls);
+
+          if (!dir_in_use && check_for_repo_containing_partial_pull (cache_dir, commit, out_location, passwd))
+            return TRUE;
+        }
+    }
+
+  return FALSE;
+}
+
+static gboolean
+handle_get_revokefs_fd (FlatpakSystemHelper   *object,
+                        GDBusMethodInvocation *invocation,
+                        GUnixFDList           *arg_fdlist,
+                        guint                  arg_flags,
+                        const gchar           *arg_ref,
+                        const gchar           *arg_commit,
+                        const gchar           *arg_installation)
+{
+  g_autoptr(FlatpakDir) system = NULL;
+  g_autoptr(GUnixFDList) fd_list = NULL;
+  g_autoptr(GError) error = NULL;
+  g_auto(GStrv) parts = NULL;
+  g_autofree gchar *src = NULL;
+  g_autofree gchar *flatpak_dir = NULL;
+  g_autofree gchar *repo_tmp = NULL;
+  g_autofree gchar *cache_dir_basename = NULL;
+  g_autofree gchar *partial_pull_repo = NULL;
+  g_autofree gchar *passwd_buf = NULL;
+  struct passwd passwd;
+  OngoingPull *new_pull;
+  uid_t uid;
+  int fd_index = -1;
+
+  g_debug ("GetRevokefsFd %u %s %s %s", arg_flags, arg_ref, arg_commit, arg_installation);
+
+  system = dir_get_system (arg_installation, get_sender_pid (invocation), &error);
+  if (system == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if ((arg_flags & ~FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_ALL) != 0)
+    {
+      g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_INVALID_ARGS,
+                                             "Unsupported flags enabled: 0x%x", (arg_flags & ~FLATPAK_HELPER_GET_REVOKEFS_FD_FLAGS_ALL));
+      return TRUE;
+    }
+
+  if (!check_for_system_helper_user (&passwd, &passwd_buf, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  if (!get_connection_uid (invocation, &uid, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  flatpak_dir = g_file_get_path (flatpak_dir_get_path (system));
+  repo_tmp = g_build_filename (flatpak_dir, "repo", "tmp", NULL);
+  parts = flatpak_decompose_ref (arg_ref, NULL);
+
+   if (check_partial_pull_availability (repo_tmp, parts[1], arg_commit, &partial_pull_repo, &passwd))
+    {
+      g_debug ("Found a partial pull for ref %s at: %s", parts[1], partial_pull_repo);
+      src = g_path_get_dirname (partial_pull_repo);
+    }
+  else
+    {
+      /* Create backing directory and let it owned by "flatpak" user */
+      cache_dir_basename = g_strdup_printf ("flatpak-cache-%s-XXXXXX", parts[1]);
+      src = g_mkdtemp_full (g_build_filename (repo_tmp, cache_dir_basename, NULL), 0755);
+      if (chown (src, passwd.pw_uid, passwd.pw_gid) == -1)
```

Comment:
When the system helper is running on the session bus as an unprivileged user during build-time tests, looking up the `struct passwd` for the system user will also have to be skipped. A typical distro workflow looks like this:

* Some autobuild server (Debian buildd, Fedora Koji, Ubuntu Launchpad, OBS, etc.)
    * install build-dependencies :warning:
    * ./configure, make
    * make check [1]
    * make install DESTDIR=...
    * pack up package from DESTDIR
* End-user system (a totally separate machine)
    * unpack package :warning:
    * run dpkg maintainer scripts, RPM scriptlets, systemd-tmpfiles, etc. to create system user [2] :warning:
    * run Flatpak
        * launch system helper :warning:

The build-time tests (`make check`) run at the point marked [1], but the system user doesn't get created until [2], so the build-time tests can't rely on the user even existing yet. They also can't create the system user themselves, because only the steps marked :warning: are typically allowed to run as root.

Reply:
In principle if the system allows you to create your own user ns, or we ran the test inside bubblewrap, you could use uidmap (which has a setuid helper, which verifies against /etc/subuid to define which auxilliary UIDs you are allowed to use) to set up a mapping of the flatpak user in the test environment to a different one outside. Not sure if unleashing bubblewrap and uidmap in a test suite is a good idea, but it's a possibility...

Reply:
With my Debian hat on, I'd prefer to be able to run as many of the tests at build-time as I can (because that's currently our only opportunity to detect regressions on non-x86_64), and our buildds can't run bubblewrap (I think it becomes confused by being run inside schroot, so the root directory isn't actually a mount point).

Reply:
I made the tests run the revokefs-fuse stuff as the user when on the session buss, so marking this resolved.